File: packages/swt3-ai-ts/src/doctor.ts
/**
 * SWT3 Doctor -- diagnostic checks for config health.
 */

import { existsSync, readFileSync, realpathSync, writeFileSync, mkdtempSync, rmSync } from "node:fs";
import { execSync } from "node:child_process";
import { resolve, dirname, join } from "node:path";
import { tmpdir } from "node:os";
import { validateSchema } from "./schema.js";

export interface DoctorCheck {
  name: string;
  status: "pass" | "warn" | "fail";
  message: string;
  tip?: string;
}

function findConfigPath(): string | null {
  for (const name of ["swt3.yaml", ".swt3.yaml"]) {
    if (existsSync(name)) return name;
  }
  return null;
}

function checkYamlFound(): DoctorCheck {
  const path = findConfigPath();
  if (path) {
    return { name: "Config file", status: "pass", message: `./${path}` };
  }
  return {
    name: "Config file",
    status: "fail",
    message: "not found",
    tip: "Run: swt3 init",
  };
}

function checkYamlValid(path: string): DoctorCheck {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const yaml = require("yaml");
    const content = readFileSync(path, "utf-8");
    const raw = yaml.parse(content);
    if (!raw || typeof raw !== "object") {
      return { name: "YAML syntax", status: "fail", message: "not a valid YAML mapping" };
    }
    return { name: "YAML syntax", status: "pass", message: "valid" };
  } catch (err) {
    return { name: "YAML syntax", status: "fail", message: (err as Error).message };
  }
}

function checkEnvVars(path: string): DoctorCheck {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const yaml = require("yaml");
    const raw = yaml.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
    const missing: string[] = [];

    if (raw.api_key_env) {
      const varName = raw.api_key_env as string;
      if (!process.env[varName]) missing.push(varName);
    }
    if (raw.signing_key_env) {
      const varName = raw.signing_key_env as string;
      if (!process.env[varName]) missing.push(varName);
    }

    if (missing.length === 0 && !raw.api_key_env && !raw.api_key) {
      return { name: "Environment", status: "warn", message: "no api_key or api_key_env configured" };
    }
    if (missing.length > 0) {
      return {
        name: "Environment",
        status: "warn",
        message: `${missing.join(", ")} not set (local mode only)`,
        tip: `export ${missing[0]}=axm_... (get a free key at https://sovereign.tenova.io/signup)`,
      };
    }
    return { name: "Environment", status: "pass", message: "all env vars resolved" };
  } catch {
    return { name: "Environment", status: "warn", message: "could not parse config" };
  }
}

function checkProfile(path: string): DoctorCheck {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const yaml = require("yaml");
    const raw = yaml.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
    const profile = raw.profile as string | undefined;
    if (!profile) {
      return { name: "Profile", status: "pass", message: "none (custom config)" };
    }
    const valid = ["eu-ai-act-high-risk", "nist-ai-rmf", "minimal"];
    if (valid.includes(profile)) {
      return { name: "Profile", status: "pass", message: profile };
    }
    return { name: "Profile", status: "fail", message: `unknown: ${profile}` };
  } catch {
    return { name: "Profile", status: "warn", message: "could not parse" };
  }
}

function checkSections(path: string): DoctorCheck {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const yaml = require("yaml");
    const raw = yaml.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
    const result = validateSchema(raw);
    if (result.valid) {
      const sections = ["policy", "trust_mesh", "hardware", "density_policy", "mcp_policy", "merkle"]
        .filter((s) => s in raw);
      return {
        name: "Sections",
        status: "pass",
        message: sections.length > 0 ? sections.join(", ") : "none configured",
      };
    }
    const first = result.errors[0];
    return {
      name: "Sections",
      status: "fail",
      message: `${first.path}: ${first.message}`,
    };
  } catch {
    return { name: "Sections", status: "warn", message: "could not validate" };
  }
}

function checkExtends(path: string): DoctorCheck {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const yaml = require("yaml");
    const raw = yaml.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
    const ext = raw.extends as string | string[] | undefined;
    if (!ext) {
      return { name: "Extends", status: "pass", message: "none" };
    }
    const files = Array.isArray(ext) ? ext : [ext];
    const configDir = dirname(resolve(path));
    const missing: string[] = [];
    for (const f of files) {
      const resolved = f.startsWith("/") ? f : join(configDir, f);
      if (!existsSync(resolved)) missing.push(f);
    }
    if (missing.length > 0) {
      return { name: "Extends", status: "fail", message: `missing: ${missing.join(", ")}` };
    }
    return { name: "Extends", status: "pass", message: `${files.length} file(s) resolved` };
  } catch {
    return { name: "Extends", status: "warn", message: "could not check" };
  }
}

function checkTpm(): DoctorCheck {
  if (process.platform !== "linux") {
    return { name: "Hardware", status: "warn", message: `${process.platform} (TPM check skipped)` };
  }
  const hasDev = existsSync("/dev/tpm0");
  let hasTools = false;
  try {
    execSync("which tpm2_pcrread", { timeout: 2_000, stdio: ["pipe", "pipe", "pipe"] });
    hasTools = true;
  } catch { /* not installed */ }

  if (hasDev && hasTools) {
    return { name: "Hardware", status: "pass", message: "/dev/tpm0 + tpm2-tools detected" };
  }
  if (hasDev) {
    return { name: "Hardware", status: "warn", message: "/dev/tpm0 detected, tpm2-tools missing",
      tip: "Install: apt install tpm2-tools" };
  }
  if (hasTools) {
    return { name: "Hardware", status: "warn", message: "tpm2-tools installed, /dev/tpm0 not detected" };
  }
  return { name: "Hardware", status: "warn", message: "/dev/tpm0 not detected" };
}

function checkRuntimeProfile(configPath: string): DoctorCheck {
  try {
    const yaml = require("yaml");
    const raw = yaml.parse(readFileSync(configPath, "utf-8"));
    const hw = raw?.hardware;
    if (!hw?.runtime_profile) {
      return { name: "Runtime Profile", status: "pass", message: "not configured (optional)" };
    }
    const rp = hw.runtime_profile;
    const fields = Object.keys(rp).filter((k: string) =>
      ["expected_topology", "min_gpu_count", "min_memory_mb", "expected_accelerator", "max_temperature_celsius", "max_power_watts"].includes(k),
    );
    return { name: "Runtime Profile", status: "pass", message: `${fields.length} constraint(s): ${fields.join(", ")}` };
  } catch {
    return { name: "Runtime Profile", status: "warn", message: "could not parse config" };
  }
}

function checkMcpConfig(): DoctorCheck {
  if (process.env.SWT3_CONFIG_FILE) {
    return { name: "MCP", status: "pass", message: `SWT3_CONFIG_FILE=${process.env.SWT3_CONFIG_FILE}` };
  }
  return {
    name: "MCP",
    status: "warn",
    message: "SWT3_CONFIG_FILE not set",
    tip: "export SWT3_CONFIG_FILE=./swt3.yaml",
  };
}

export function runDoctorChecks(configPath?: string): DoctorCheck[] {
  const checks: DoctorCheck[] = [];

  let path: string | null;
  if (configPath) {
    if (existsSync(configPath)) {
      checks.push({ name: "Config file", status: "pass", message: configPath });
      path = configPath;
    } else {
      checks.push({ name: "Config file", status: "fail", message: `not found: ${configPath}`, tip: "Run: swt3 init" });
      return checks;
    }
  } else {
    const yamlCheck = checkYamlFound();
    checks.push(yamlCheck);
    path = yamlCheck.status === "pass" ? yamlCheck.message.replace("./", "") : null;
  }

  if (!path) {
    return checks;
  }

  checks.push(checkYamlValid(path));
  checks.push(checkEnvVars(path));
  checks.push(checkProfile(path));
  checks.push(checkSections(path));
  checks.push(checkExtends(path));
  checks.push(checkTpm());
  checks.push(checkRuntimeProfile(path));
  checks.push(checkMcpConfig());

  return checks;
}

export function printDoctorResults(checks: DoctorCheck[], json: boolean = false): void {
  if (json) {
    console.log(JSON.stringify(checks, null, 2));
    return;
  }

  const VERSION = "0.5.2";
  console.log(`\n  SWT3 Doctor v${VERSION}\n`);

  const icons = { pass: "\x1b[32m[PASS]\x1b[0m", warn: "\x1b[33m[WARN]\x1b[0m", fail: "\x1b[31m[FAIL]\x1b[0m" };

  for (const check of checks) {
    const icon = icons[check.status];
    console.log(`  ${icon} ${check.name}: ${check.message}`);
    if (check.tip) {
      console.log(`         Tip: ${check.tip}`);
    }
  }

  const pass = checks.filter((c) => c.status === "pass").length;
  const warn = checks.filter((c) => c.status === "warn").length;
  const fail = checks.filter((c) => c.status === "fail").length;
  console.log(`\n  ${pass} passed, ${warn} warnings, ${fail} failures\n`);

  console.log("  \x1b[33mNew in v0.5.2:\x1b[0m Trust Mesh -- agents verify each other before exchanging data.");
  console.log("  Configure: \x1b[36mswt3 init --profile eu-ai-act-high-risk\x1b[0m");
  console.log("  Docs: https://www.npmjs.com/package/@tenova/swt3-mcp\n");
}

// ── Friction Test ───────────────────────────────────────────────────────

export interface FrictionStep {
  name: string;
  status: "pass" | "fail";
  durationMs: number;
  error?: string;
}

export async function runFrictionTest(): Promise<FrictionStep[]> {
  const steps: FrictionStep[] = [];
  const tmpDir = mkdtempSync(join(tmpdir(), "swt3-friction-"));
  const configPath = join(tmpDir, "swt3.yaml");

  function step(name: string, fn: () => void): void {
    const start = performance.now();
    try {
      fn();
      steps.push({ name, status: "pass", durationMs: Math.round(performance.now() - start) });
    } catch (err: any) {
      steps.push({ name, status: "fail", durationMs: Math.round(performance.now() - start), error: err.message });
    }
  }

  // 1. Config discovery
  step("Config discovery", () => {
    writeFileSync(configPath, [
      "clearing_level: 1",
      "tenant_id: FRICTION_TEST",
      "agent_id: test-agent",
      "api_key: test_key_friction",
      "mcp_policy:",
      "  tool_blocklist: [\"dangerous_tool\"]",
      "  max_tokens_per_session: 500",
      "  fail_secure: true",
    ].join("\n"), "utf-8");
    if (!existsSync(configPath)) throw new Error("Config file not created");
  });

  // 2. fromConfig loads
  step("fromConfig() loads", () => {
    const { loadFullConfig } = require("./config.js");
    const loaded = loadFullConfig(configPath);
    if (!loaded.mcpPolicy) throw new Error("mcpPolicy not parsed");
    if (!loaded.mcpPolicy.toolBlocklist?.length) throw new Error("toolBlocklist empty");
  });

  // 3. ChainEnforcer created
  step("ChainEnforcer created", () => {
    const { ChainEnforcer } = require("./witness.js");
    const enforcer = new ChainEnforcer({
      witnessedTools: [], exemptTools: [], requireTrustLevel: 0,
      autoWitness: true, blockOnFailure: false,
      toolBlocklist: ["dangerous_tool"], failSecure: true,
    });
    const result = enforcer.check("safe_tool");
    if (result !== null) throw new Error("Safe tool was blocked");
  });

  // 4. wrapTool blocks on violation
  step("Blocklist enforcement", () => {
    const { ChainEnforcer, PolicyViolationError } = require("./witness.js");
    const enforcer = new ChainEnforcer({
      witnessedTools: [], exemptTools: [], requireTrustLevel: 0,
      autoWitness: true, blockOnFailure: false,
      toolBlocklist: ["dangerous_tool"], failSecure: true,
    });
    const violation = enforcer.check("dangerous_tool");
    if (!violation) throw new Error("Blocklist did not trigger");
    if (violation.action !== "blocked") throw new Error("Action should be blocked");
  });

  // 5. Token budget enforcement
  step("Token budget enforcement", () => {
    const { ChainEnforcer } = require("./witness.js");
    const enforcer = new ChainEnforcer({
      witnessedTools: [], exemptTools: [], requireTrustLevel: 0,
      autoWitness: true, blockOnFailure: false,
      maxTokensPerSession: 100, failSecure: true,
    });
    enforcer.recordTokens(150);
    const violation = enforcer.check("any_tool");
    if (!violation) throw new Error("Token budget did not trigger");
    if (violation.rule !== "token_budget") throw new Error(`Wrong rule: ${violation.rule}`);
  });

  // 6. Audit report generation
  step("Audit report generation", () => {
    const { ChainMonitorExporter } = require("./exporters/chain-monitor.js");
    const exporter = new ChainMonitorExporter({ walDir: tmpDir, tenantId: "FRICTION_TEST" });
    const html = exporter.exportHtml();
    if (!html.includes("SWT3 Exploit Chain Monitor")) throw new Error("HTML missing title");
    if (!html.includes("Self-Signed")) throw new Error("HTML missing watermark");
  });

  // 7. Error message clarity
  step("Error message clarity", () => {
    const { PolicyViolationError } = require("./witness.js");
    const violation = {
      rule: "blocklist", toolName: "dangerous_tool", action: "blocked",
      reason: 'Tool "dangerous_tool" is on the blocklist', timestamp: Date.now(),
    };
    const err = new PolicyViolationError(violation);
    if (!err.message.includes("dangerous_tool")) throw new Error("Error missing tool name");
    if (!err.message.includes("blocklist")) throw new Error("Error missing reason");
  });

  // Cleanup
  try { rmSync(tmpDir, { recursive: true, force: true }); } catch { /* ignore */ }

  return steps;
}

export function printFrictionResults(steps: FrictionStep[]): void {
  const VERSION = "0.5.3";
  console.log(`\n  SWT3 Friction Test v${VERSION}\n`);

  for (let i = 0; i < steps.length; i++) {
    const s = steps[i];
    const icon = s.status === "pass" ? "\x1b[32mPASS\x1b[0m" : "\x1b[31mFAIL\x1b[0m";
    const dots = ".".repeat(Math.max(1, 36 - s.name.length));
    console.log(`  ${i + 1}. ${s.name} ${dots} ${icon}  (${s.durationMs}ms)`);
    if (s.error) {
      console.log(`     \x1b[31m${s.error}\x1b[0m`);
    }
  }

  const passed = steps.filter((s) => s.status === "pass").length;
  const total = steps.length;
  const totalMs = steps.reduce((sum, s) => sum + s.durationMs, 0);
  console.log(`\n  ${passed}/${total} passed. Total: ${totalMs}ms.\n`);
}

File: packages/swt3-ai/swt3_ai/doctor.py
"""SWT3 Doctor -- diagnostic checks for config health."""

from __future__ import annotations

import json
import os
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import List, Optional

from .schema import validate_schema

VERSION = "0.5.2"


@dataclass
class DoctorCheck:
    name: str
    status: str  # "pass" | "warn" | "fail"
    message: str
    tip: str = ""


def _find_config_path() -> Optional[str]:
    for name in ("swt3.yaml", ".swt3.yaml"):
        if Path(name).is_file():
            return name
    return None


def _check_yaml_found() -> DoctorCheck:
    path = _find_config_path()
    if path:
        return DoctorCheck("Config file", "pass", f"./{path}")
    return DoctorCheck("Config file", "fail", "not found", tip="Run: swt3 init")


def _check_yaml_valid(path: str) -> DoctorCheck:
    try:
        import yaml
        with open(path, "r") as f:
            raw = yaml.safe_load(f)
        if not isinstance(raw, dict):
            return DoctorCheck("YAML syntax", "fail", "not a valid YAML mapping")
        return DoctorCheck("YAML syntax", "pass", "valid")
    except Exception as e:
        return DoctorCheck("YAML syntax", "fail", str(e))


def _check_env_vars(path: str) -> DoctorCheck:
    try:
        import yaml
        with open(path, "r") as f:
            raw = yaml.safe_load(f)
        missing = []
        if raw.get("api_key_env"):
            var_name = raw["api_key_env"]
            if not os.environ.get(var_name):
                missing.append(var_name)
        if raw.get("signing_key_env"):
            var_name = raw["signing_key_env"]
            if not os.environ.get(var_name):
                missing.append(var_name)

        if not missing and not raw.get("api_key_env") and not raw.get("api_key"):
            return DoctorCheck("Environment", "pass", "local mode (no api_key configured)")
        if missing:
            return DoctorCheck(
                "Environment", "warn",
                f"{', '.join(missing)} not set",
                tip=f"export {missing[0]}=axm_... (get a free key at https://sovereign.tenova.io/signup)",
            )
        return DoctorCheck("Environment", "pass", "all env vars resolved")
    except Exception:
        return DoctorCheck("Environment", "warn", "could not parse config")


def _check_profile(path: str) -> DoctorCheck:
    try:
        import yaml
        with open(path, "r") as f:
            raw = yaml.safe_load(f)
        profile = raw.get("profile")
        if not profile:
            return DoctorCheck("Profile", "pass", "none (custom config)")
        valid = {"eu-ai-act-high-risk", "nist-ai-rmf", "minimal"}
        if profile in valid:
            return DoctorCheck("Profile", "pass", profile)
        return DoctorCheck("Profile", "fail", f"unknown: {profile}")
    except Exception:
        return DoctorCheck("Profile", "warn", "could not parse")


def _check_sections(path: str) -> DoctorCheck:
    try:
        import yaml
        with open(path, "r") as f:
            raw = yaml.safe_load(f)
        result = validate_schema(raw)
        if result.valid:
            sections = [
                s for s in ("policy", "trust_mesh", "hardware", "density_policy", "mcp_policy", "merkle")
                if s in raw
            ]
            return DoctorCheck(
                "Sections", "pass",
                ", ".join(sections) if sections else "none configured",
            )
        first = result.errors[0]
        return DoctorCheck("Sections", "fail", f"{first.path}: {first.message}")
    except Exception:
        return DoctorCheck("Sections", "warn", "could not validate")


def _check_extends(path: str) -> DoctorCheck:
    try:
        import yaml
        with open(path, "r") as f:
            raw = yaml.safe_load(f)
        ext = raw.get("extends")
        if not ext:
            return DoctorCheck("Extends", "pass", "none")
        files = ext if isinstance(ext, list) else [ext]
        config_dir = Path(path).resolve().parent
        missing = []
        for f in files:
            resolved = Path(f) if Path(f).is_absolute() else config_dir / f
            if not resolved.is_file():
                missing.append(f)
        if missing:
            return DoctorCheck("Extends", "fail", f"missing: {', '.join(missing)}")
        return DoctorCheck("Extends", "pass", f"{len(files)} file(s) resolved")
    except Exception:
        return DoctorCheck("Extends", "warn", "could not check")


def _check_tpm(config_path: Optional[str] = None) -> DoctorCheck:
    import shutil

    require_attestation = False
    if config_path:
        try:
            import yaml as _yaml
            raw = _yaml.safe_load(Path(config_path).read_text())
            hw = raw.get("hardware") if isinstance(raw, dict) else None
            require_attestation = isinstance(hw, dict) and hw.get("require_attestation") is True
        except Exception:
            pass

    if sys.platform != "linux":
        if require_attestation:
            return DoctorCheck("Hardware", "warn", f"{sys.platform} (TPM required but not Linux)")
        return DoctorCheck("Hardware", "pass", f"{sys.platform} (TPM not required)")

    has_dev = Path("/dev/tpm0").exists()
    has_tools = shutil.which("tpm2_pcrread") is not None

    if has_dev and has_tools:
        return DoctorCheck("Hardware", "pass", "/dev/tpm0 + tpm2-tools detected")
    if not require_attestation:
        return DoctorCheck("Hardware", "pass", "not required")
    if has_dev:
        return DoctorCheck("Hardware", "warn", "/dev/tpm0 detected, tpm2-tools missing")
    return DoctorCheck("Hardware", "warn", "TPM required but /dev/tpm0 not detected")


def _check_runtime_profile(config_path: str) -> DoctorCheck:
    try:
        import yaml as _yaml
        raw = _yaml.safe_load(Path(config_path).read_text())
        hw = raw.get("hardware") if isinstance(raw, dict) else None
        if not isinstance(hw, dict) or "runtime_profile" not in hw:
            return DoctorCheck("Runtime Profile", "pass", "not configured (optional)")
        rp = hw["runtime_profile"]
        valid = {"expected_topology", "min_gpu_count", "min_memory_mb", "expected_accelerator", "max_temperature_celsius", "max_power_watts"}
        fields = [k for k in rp if k in valid]
        return DoctorCheck("Runtime Profile", "pass", f"{len(fields)} constraint(s): {', '.join(fields)}")
    except Exception:
        return DoctorCheck("Runtime Profile", "warn", "could not parse config")


def _check_mcp_config() -> DoctorCheck:
    val = os.environ.get("SWT3_CONFIG_FILE")
    if val:
        return DoctorCheck("MCP", "pass", f"SWT3_CONFIG_FILE={val}")
    return DoctorCheck("MCP", "pass", "not configured (optional)")


def run_doctor_checks(config_path: Optional[str] = None) -> List[DoctorCheck]:
    checks: List[DoctorCheck] = []

    if config_path:
        if Path(config_path).is_file():
            checks.append(DoctorCheck("Config file", "pass", config_path))
            path: Optional[str] = config_path
        else:
            checks.append(DoctorCheck("Config file", "fail", f"not found: {config_path}", tip="Run: swt3 init"))
            return checks
    else:
        yaml_check = _check_yaml_found()
        checks.append(yaml_check)
        path = yaml_check.message.replace("./", "") if yaml_check.status == "pass" else None

    if not path:
        return checks

    checks.append(_check_yaml_valid(path))
    checks.append(_check_env_vars(path))
    checks.append(_check_profile(path))
    checks.append(_check_sections(path))
    checks.append(_check_extends(path))
    checks.append(_check_tpm(path))
    checks.append(_check_runtime_profile(path))
    checks.append(_check_mcp_config())

    return checks


def print_doctor_results(checks: List[DoctorCheck], use_json: bool = False, ci_mode: bool = False) -> None:
    if use_json:
        import json as json_mod
        print(json_mod.dumps([{"name": c.name, "status": c.status, "message": c.message, "tip": c.tip} for c in checks], indent=2))
        return

    p = sum(1 for c in checks if c.status == "pass")
    w = sum(1 for c in checks if c.status == "warn")
    fl = sum(1 for c in checks if c.status == "fail")

    if ci_mode:
        for check in checks:
            tag = check.status.upper()
            print(f"[{tag}] {check.name}: {check.message}")
            if check.tip:
                print(f"  Tip: {check.tip}")
        print(f"swt3-doctor: {len(checks)} checks, {p} pass, {w} warn, {fl} fail")
        return

    print(f"\n  SWT3 Doctor v{VERSION}\n")

    icons = {"pass": "\033[32m[PASS]\033[0m", "warn": "\033[33m[WARN]\033[0m", "fail": "\033[31m[FAIL]\033[0m"}

    for check in checks:
        icon = icons.get(check.status, "[????]")
        print(f"  {icon} {check.name}: {check.message}")
        if check.tip:
            print(f"         Tip: {check.tip}")

    print(f"\n  {p} passed, {w} warnings, {fl} failures\n")

    print("  \033[33mNew in v0.5.2:\033[0m Trust Mesh -- agents verify each other before exchanging data.")
    print("  Configure: \033[36mswt3 init --profile eu-ai-act-high-risk\033[0m")
    print("  Docs: https://www.npmjs.com/package/@tenova/swt3-mcp\n")


# ── Friction Test ───────────────────────────────────────────────────────

@dataclass
class FrictionStep:
    name: str = ""
    status: str = "pass"
    duration_ms: int = 0
    error: Optional[str] = None


def run_friction_test() -> List[FrictionStep]:
    import tempfile
    import time
    import shutil

    steps: List[FrictionStep] = []
    tmp_dir = tempfile.mkdtemp(prefix="swt3-friction-")
    config_path = os.path.join(tmp_dir, "swt3.yaml")

    def step(name: str, fn):
        start = time.monotonic()
        try:
            fn()
            steps.append(FrictionStep(name=name, status="pass",
                duration_ms=int((time.monotonic() - start) * 1000)))
        except Exception as e:
            steps.append(FrictionStep(name=name, status="fail",
                duration_ms=int((time.monotonic() - start) * 1000), error=str(e)))

    # 1. Config discovery
    def _config():
        with open(config_path, "w") as f:
            f.write("\n".join([
                "clearing_level: 1",
                "tenant_id: FRICTION_TEST",
                "agent_id: test-agent",
                "api_key: test_key_friction",
                "mcp_policy:",
                "  tool_blocklist: [\"dangerous_tool\"]",
                "  max_tokens_per_session: 500",
                "  fail_secure: true",
            ]))
        if not os.path.exists(config_path):
            raise ValueError("Config file not created")
    step("Config discovery", _config)

    # 2. fromConfig loads
    def _load():
        from .config import load_full_config
        loaded = load_full_config(config_path)
        if not loaded.mcp_policy:
            raise ValueError("mcp_policy not parsed")
        if not loaded.mcp_policy.tool_blocklist:
            raise ValueError("tool_blocklist empty")
    step("from_config() loads", _load)

    # 3. ChainEnforcer created
    def _enforcer():
        from .witness import ChainEnforcer
        from .types import McpPolicyConfig
        enforcer = ChainEnforcer(McpPolicyConfig(
            tool_blocklist=["dangerous_tool"], fail_secure=True,
        ))
        result = enforcer.check("safe_tool")
        if result is not None:
            raise ValueError("Safe tool was blocked")
    step("ChainEnforcer created", _enforcer)

    # 4. Blocklist enforcement
    def _blocklist():
        from .witness import ChainEnforcer
        from .types import McpPolicyConfig
        enforcer = ChainEnforcer(McpPolicyConfig(
            tool_blocklist=["dangerous_tool"], fail_secure=True,
        ))
        v = enforcer.check("dangerous_tool")
        if not v:
            raise ValueError("Blocklist did not trigger")
        if v.action != "blocked":
            raise ValueError("Action should be blocked")
    step("Blocklist enforcement", _blocklist)

    # 5. Token budget enforcement
    def _tokens():
        from .witness import ChainEnforcer
        from .types import McpPolicyConfig
        enforcer = ChainEnforcer(McpPolicyConfig(
            max_tokens_per_session=100, fail_secure=True,
        ))
        enforcer.record_tokens(150)
        v = enforcer.check("any_tool")
        if not v:
            raise ValueError("Token budget did not trigger")
        if v.rule != "token_budget":
            raise ValueError(f"Wrong rule: {v.rule}")
    step("Token budget enforcement", _tokens)

    # 6. Audit report generation
    def _audit():
        from .exporters.chain_monitor import ChainMonitorExporter
        exporter = ChainMonitorExporter(wal_dir=tmp_dir, tenant_id="FRICTION_TEST")
        html_out = exporter.export_html()
        if "SWT3 Exploit Chain Monitor" not in html_out:
            raise ValueError("HTML missing title")
        if "Self-Signed" not in html_out:
            raise ValueError("HTML missing watermark")
    step("Audit report generation", _audit)

    # 7. Error message clarity
    def _errors():
        from .witness import PolicyViolationError
        from .types import ChainPolicyViolation
        v = ChainPolicyViolation(
            rule="blocklist", tool_name="dangerous_tool", action="blocked",
            reason='Tool "dangerous_tool" is on the blocklist', timestamp=0.0,
        )
        err = PolicyViolationError(v)
        if "dangerous_tool" not in str(err):
            raise ValueError("Error missing tool name")
        if "blocklist" not in str(err):
            raise ValueError("Error missing reason")
    step("Error message clarity", _errors)

    try:
        shutil.rmtree(tmp_dir, ignore_errors=True)
    except Exception:
        pass

    return steps


def print_friction_results(steps: List[FrictionStep]) -> None:
    print(f"\n  SWT3 Friction Test v{VERSION}\n")
    for i, s in enumerate(steps):
        icon = "\033[32mPASS\033[0m" if s.status == "pass" else "\033[31mFAIL\033[0m"
        dots = "." * max(1, 36 - len(s.name))
        print(f"  {i + 1}. {s.name} {dots} {icon}  ({s.duration_ms}ms)")
        if s.error:
            print(f"     \033[31m{s.error}\033[0m")
    passed = sum(1 for s in steps if s.status == "pass")
    total_ms = sum(s.duration_ms for s in steps)
    print(f"\n  {passed}/{len(steps)} passed. Total: {total_ms}ms.\n")
